### PR TITLE
Implement basic categorization for libraries in dartdoc

### LINKED ISSUE #1353: Request to group concepts together, across libraries
Related to https://github.com/flutter/flutter/issues/8906

We'd like to:

* somehow say "these concepts are all part of the same concept/category/grouping"
* somehow say "this is the homepage for that concept"

### LINKED ISSUE #1610: Support custom categories for libraries within a package – and the SDK
Allow organizing the libraries (on the left side of nav) into custom categories

Blocking https://github.com/dart-lang/sdk/issues/32297

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #1610.

Also, this is the beginning of a longer chain to implement #1353.

This allows you to mark libraries with `{@category Foo}` and have libraries appear in a category in the sidebar and in the main package view.  Uncategorized libraries appear at the top.  An example:

https://user-images.githubusercontent.com/14116827/37685847-3f34c478-2c52-11e8-82b9-6b1485a5a05d.png

This PR also eliminates the "use-categories" flag.  The "old style" meaning of categories meaning packages is now gone.  The flag is now ignored and Dartdoc decides whether to organize libraries into packages based on whether there are multiple packages in the requested set of libraries.

### CODE REVIEW

**Comment by jcollins-g** [comment]
This PR also has some implications for #739 and #1454 -- strengthening the Package class and driving more through it will eventually enable those.

File: README.md [hunk 1]
```diff
@@ -95,6 +95,34 @@ authoring doc comments for Dart with `dartdoc`.
 
 ## Advanced features
 
+### dartdoc_options.yaml
+
+Creating a file named dartdoc_options.yaml at the top of your package can change how Dartdoc
+generates docs.  
+
+```yaml
+dartdoc:
```
Comment (kevmoo) [inline README.md]: Why have a `dartdoc` top key in `dartdoc_options.yaml`?

Couldn't `categoryOrder` just be at the top?

File: README.md [hunk 2]
```diff
@@ -114,6 +142,9 @@ and then you can insert it via `{@macro template_name}`, like
 /// More comments
 ```
 
+
```
Comment (kevmoo) [inline README.md]: extra whitespace?

Comment (jcollins-g) [reply README.md]: done.

File: README.md (hunk 1 above)
Comment (jcollins-g) [reply README.md]: This was paralleling with `analysis_options.yaml`, which has analyzer and linter subsections.  There might be an opportunity for sections for other tools using dartdoc libraries, such as the still vaporware API comparison gadget.

**Comment by kevmoo** [comment]
ack

**Pull request merged**